packages/react-combobox/src/utils/useOptionCollection.ts

```diff
@@ -0,0 +1,90 @@
+import * as React from 'react';
+import { OptionProps } from '../components/Option/Option.types';
+import type { OptionCollectionState, OptionData, OptionValue } from './OptionCollection.types';
+
+/**
+ * Takes React children and returns an array of Option keys in order, and a cloned set of processed children
+ */
+function getValidOptions(children: React.ReactNode): { keys: string[]; children: React.ReactNode } {
+  const keys: string[] = [];
+
+  const clonedChildren = React.Children.map(children, (child, index) => {
+    if (React.isValidElement(child) && typeof child.type === 'object') {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const { fluentComponentType } = child.type as any;
+
+      // if the child is an option, add its key to the array
+      if (fluentComponentType === 'Option') {
+        const optionKey = child.props.fluentKey || child.key || child.props.id || `${index}`;
```

Comment:
This is really **dangerous** pattern, we should find a better option.

This breaks composition and makes the component dependant on JSX structure. It kinda allows to use JSX children, but breaks on any React composition pattern:

```tsx
function App() {
  return (
    <>
      {/* 💥 This does not work */}
      <Combobox>
        <MyOptions />
      </Combobox>
      {/* 💥 This does not work */}
      <Combobox>
        <CustomOption /> {/* ⬅️ wraps "Option" */}
      </Combobox>
    </>
  );
}
```

Using static properties (_I mean `fluentComponentType`_) as identifiers also breaks composition patterns:

```tsx
const MemoOption = React.memo(Option)

function App() {
  return (
      <Combobox>
      {/* ✅ This works */}
        <Option />
      {/* 💥 This does not work */}
        <CustomOption />
      </Combobox>
    </>
  );
}
```

Reply:
I agree - wrapping of components is something we need to account for as it is commonly used for layering the application, for example for fetching the data

Reply:
https://github.com/microsoft/fluentui/issues/22055